craft_parts/plugins/python_plugin.py

```diff
@@ -154,44 +154,91 @@ def get_build_commands(self) -> List[str]:
-                        python_path="$(realpath --strip --relative-to="${{interp_dir}}" \\
-                                "${{python_path}}")"
-                        break
+                # look for a provisioned python interpreter
+                opts_state="$(set +o|grep errexit)"
+                set +e
+                install_dir="{self._part_info.part_install_dir}/usr/bin"
+                stage_dir="{self._part_info.stage_dir}/usr/bin"
+                payload_python=$(find "$install_dir" "$stage_dir" -type f -executable -name "python3.*" -print -quit 2>/dev/null)
+
+                if [ -n "$payload_python" ]; then
+                    # We found a provisioned interpreter, use it.
+                    installed_python="${{payload_python##{self._part_info.part_install_dir}}}"
+                    if [ "$installed_python" = "$payload_python" ]; then
+                        # Found a staged interpreter.
+                        symlink_target="..${{payload_python##{self._part_info.stage_dir}}}"
+                    else
+                        # The interpreter was installed but not staged yet.
+                        symlink_target="..$installed_python"
                     fi
-                done
-                echo "${{python_path}}"
-                eval "${{opts_state}}"
-            }}
+                else
+                    # Otherwise use what _get_system_python_interpreter() told us.
+                    symlink_target="{python_interpreter}"
+                fi
+
+                if [ -z "$symlink_target" ]; then
+                    echo "No suitable Python interpreter found, giving up."
+                    exit 1
+                fi
 
-            python_path="$(determine_link_target)"
-            ln -sf "${{python_path}}" "${{PARTS_PYTHON_VENV_INTERP_PATH}}"
-            """
+                eval "${{opts_state}}"
+                ln -sf "${{symlink_target}}" "${{PARTS_PYTHON_VENV_INTERP_PATH}}"
+                """
             )
         )
 
         return build_commands
+
+    def _should_remove_symlinks(self) -> bool:
+        """Configure executables symlink removal.
+
+        This method can be overridden by application-specific subclasses to control
+        whether symlinks in the virtual environment should be removed. Default is
+        False.  If True, the venv-created symlinks to python* in bin/ will be
+        removed and will not be recreated.
+        """
+        return False
+
+    def _get_system_python_interpreter(self) -> Optional[str]:
+        """Obtain the path to the system-provided python interpreter.
+
+        This method can be overridden by application-specific subclasses. It
+        returns the path to the Python that bin/python3 should be symlinked to
+        if Python is not part of the payload.
+        """
+        return '$(readlink -f "$(which "${PARTS_PYTHON_INTERPRETER}")")'
+
+    def _get_script_interpreter(self) -> str:
```

Comment:
I'm wondering if we should leave this part specifically (the shebang fix) alone for now. During our last meeting I showed that ugly hack that works for installed snaps but not during the build, so this method is either insufficient or it's going to get way uglier.
If I remember correctly the suggestion was made to add full-on scripts that do the full lookup of the correctly Python to use. So for example, when fixing "black" we would rename it to (say) "_black" and then add new "black" file as a shell script that calls "_black". Of course, this strategy still needs to be validated/prototyped.
what do you think? @cmatsuoka @sergiusens

Reply:
We could leave the method there and not override it on the Snapcraft-specific plugin for now as it doesn't seem to be causing immediate problems. We can work on it again after reexamining the scenario and possible alternatives.

Reply:
Alright, if you're fine with possibly re-working this again in the future then I'm fine with it too